app/src/main/java/com/example/plantsapp/data/firebase/repository/FirebasePlantsRepository.kt

```diff
@@ -0,0 +1,94 @@
+package com.example.plantsapp.data.firebase.repository
+
+import com.example.plantsapp.data.firebase.entity.FirebasePlant
+import com.example.plantsapp.domain.model.Plant
+import com.example.plantsapp.domain.repository.PlantsRepository
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.ktx.toObject
+import com.google.firebase.firestore.ktx.toObjects
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.tasks.await
+import timber.log.Timber
+import javax.inject.Inject
+
+class FirebasePlantsRepository @Inject constructor(
+    private val firestore: FirebaseFirestore
+) : PlantsRepository {
+
+    override fun observePlants(): Flow<List<Plant>> {
+        val flowListPlants = MutableSharedFlow<List<Plant>>()
+        firestore.collection(KEY_COLLECTION_PLANTS)
+            .addSnapshotListener { value, error ->
+                if (error != null) {
+                    Timber.e("Listen failed..", error)
+                    return@addSnapshotListener
+                }
+
+                value?.let { documents ->
+                    // TODO change GlobalScope to some other scope
+                    GlobalScope.launch {
+                        val plants = documents
+                            .toObjects<FirebasePlant>()
+                            .map { it.toPlant() }
+
+                        flowListPlants.emit(plants)
+                    }
+                }
+            }
+
+        return flowListPlants
+    }
+
+    override suspend fun fetchPlants(): List<Plant> {
+        return firestore.collection(KEY_COLLECTION_PLANTS)
+            .get()
+            .await()
+            .toObjects<FirebasePlant>()
+            .map { it.toPlant() }
+    }
+
+    override suspend fun addPlant(plant: Plant) {
+        // TODO add plantPicture to storage in Firebase and use its address here
+        firestore.collection(KEY_COLLECTION_PLANTS)
+            .document(plant.name.value)
+            .set(FirebasePlant.from(plant))
```

Comment:
You can move part where you get collection and document into some private function, so in each method you will use only needed thing